File: kotlin-in-action/src/tutorials/lesson25/Main.java
package tutorials.lesson25;

import java.io.*;
import java.nio.file.Files;

/**
 * @author kumushuoshuo
 * @github https://github.com/chiclaim/
 */
public class Main {

    private void printFile() {
        BufferedReader br = null;
        try {
            br = new BufferedReader(new FileReader("test.txt"));
            String line;
            while ((line = br.readLine()) != null) {
                System.out.println(line);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (br != null) {
                try {
                    br.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private void printFile2() {
        // try-with-resource
        try (BufferedReader br = new BufferedReader(new FileReader("test.txt"))) {
            String line;
            while ((line = br.readLine()) != null) {
                System.out.println(line);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void copy() {
        InputStream is = null;
        OutputStream os = null;
        try {
            is = new FileInputStream("D:\\imagetest\\desk.jpg");
            os = new FileOutputStream("E:\\desk1.jpg");
            byte[] buffer = new byte[1024];
            int readNum = 0;
            int a = 0;
            while ((readNum = is.read(buffer, 0, buffer.length)) != -1) {
                System.out.println(a++);
                os.write(buffer, 0, readNum);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (os != null) {
                try {
                    os.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (is != null) {
                try {
                    is.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }


}

File: kotlin-in-action/src/main/java/tutorials/lesson25/Main.kt
package tutorials.lesson25

import java.io.File

/**
 *
 * @author kumushuoshuo
 * @github https://github.com/chiclaim/
 */


val file = File("kotlin-in-action/src/tutorials/lesson25/file.txt")

fun testWrite() {
    file.writeBytes("kumushuoshuo writeBytes\n".toByteArray())
    //file.writeText("kumushuoshuo writeText\n")
    file.appendBytes("kumushuoshuo appendBytes\n".toByteArray())
    file.appendText("kumushuoshuo appendText\n")
//    file.outputStream().use {
//        it.write("kumushuoshuo outputStream\n".toByteArray())
//    }
//    file.printWriter().use {
//        it.println("kumushuoshuo printWriter")
//    }
//    file.bufferedWriter().use {
//        it.write("kumushuoshuo")
//    }

}

fun testRead() {
    println("forEachLine-------------------->")
    file.forEachLine{
        println(it)
    }
    println("forEachBlock-------------------->")
    file.forEachBlock { buffer: ByteArray, _: Int ->
        print(buffer.decodeToString())
    }
    println("inputStream-------------------->")
    file.inputStream().use {
        var value: Int = it.read()
        while (value != -1) {
            print(value.toChar())
            value = it.read()
        }
    }
    println("inputStream buffer-------------------->")
    file.inputStream().use {
        val byteArray = ByteArray(1024)
        var value: Int = it.read(byteArray,0,byteArray.size)
        while (value != -1) {
            print(byteArray.decodeToString())
            value = it.read(byteArray,0,byteArray.size)
        }
    }
    println("bufferedReader-------------------->")
    file.bufferedReader().use {
        it.readLines().forEach(::println)
    }


}

fun testOther(){
    // 文件后缀
    println(file.extension)

    // 文件名
    println(file.nameWithoutExtension)

    // 遍历文件夹
    file.parentFile.parentFile
        //.walk()// 先访问文件夹，再访问它文件夹里的内容
        .walkBottomUp()// 先访问文件，再访问它文件夹
        .maxDepth(2) // 访问深度
        //.filter {it.extension.equals("txt",true)  }
        .forEach(::println)

    // 文件拷贝
    file.copyTo(File(file.parent,"copy.txt"), overwrite = true)

    // 文件夹拷贝
    file.parentFile.copyRecursively(File(file.parentFile.parentFile.parentFile.parent,"copyRecursively"),true)
}


fun main() {
    //testWrite()
    //testRead()
    testOther()
}